solr/core/src/java/org/apache/solr/search/neural/KnnQParser.java

```diff
@@ -80,44 +82,60 @@ public Query parse() {
     float[] parsedVectorToSearch = parseVector(vectorToSearch, denseVectorType.getDimension());
 
     return denseVectorType.getKnnVectorQuery(
-        schemaField.getName(), parsedVectorToSearch, topK, getFilterQuery());
+        schemaField.getName(), parsedVectorToSearch, topK, buildPreFilter());
   }
 
-  private Query getFilterQuery() throws SolrException {
-    if (!isFilter()) {
+  private Query buildPreFilter() throws SolrException {
+    boolean isSubQuery = recurseCount != 0;
+    if (!isFilter() && !isSubQuery) {
       String[] filterQueries = req.getParams().getParams(CommonParams.FQ);
       if (filterQueries != null && filterQueries.length != 0) {
-        List<Query> filters;
+        List<Query> preFilters;
 
         try {
-          filters = QueryUtils.parseFilterQueries(req, true);
+          preFilters = acceptPreFiltersOnly(QueryUtils.parseFilterQueries(req, true));
         } catch (SyntaxError e) {
           throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, e);
         }
 
-        if (filters.size() == 1) {
-          return filters.get(0);
+        if (preFilters.size() == 0) {
+          return null;
+        } else if (preFilters.size() == 1) {
+          return preFilters.get(0);
+        } else {
+          BooleanQuery.Builder builder = new BooleanQuery.Builder();
+          for (Query query : preFilters) {
+            builder.add(query, BooleanClause.Occur.FILTER);
+          }
+          return builder.build();
         }
-
-        BooleanQuery.Builder builder = new BooleanQuery.Builder();
-        for (Query query : filters) {
-          builder.add(query, BooleanClause.Occur.FILTER);
-        }
-
-        return builder.build();
       }
     }
-
     return null;
   }
 
+  private List<Query> acceptPreFiltersOnly(List<Query> filters) {
```

Comment:
This method looks suspicious.  The cost is merely a hint; it isn't explicitly an indicator about being a post-filter or not.  Maybe I'm missing the point?

Reply:
Look at the javadocs for PostFilter.  Basically `implements PostFilter` and `score >= 100`.

Reply:
Thanks for the useful insight David! Is it any better now?

Reply:
Can we name this method `excludePostFilters`?  I've never heard of "pre-filter" before.  Any part of a query that isn't a PostFilter is not necessarily executed in a defined order (is not "pre"); it's a combined query tree with interesting leap-frogging behavior based on various costs.

The logic appears correct but I'd like to recommend a simplification to make it even easier to read.  I suggest adding a static utility method on PostFilter called `boolean runAsPostFilter(Query q)`.  You could then simply make this a method reference.  Perhaps use this elsewhere (e.g. in SolrIndexSearcher.getProcessedFilter).